view/frontend/web/js/utils/when-defined.js

```diff
@@ -29,45 +29,47 @@ define([], function () {
-    function whenDefined(object, property, callback, key = null) {
-        if (object.hasOwnProperty(property)) {
+    function whenDefined (object, property, callback, key) {
+        if (object.hasOwnProperty(property) && typeof object[property] !== 'undefined') {
             callback();
         } else {
             var overloadedPropertyName = '_' + property;
             if (!whenDefinedCallbacks.has(object)) {
                 whenDefinedCallbacks.set(object, new Map([]));
             }
             if (!whenDefinedCallbacks.get(object).has(property)) {
-                whenDefinedCallbacks.get(object).set(property, []);
-            }
-            var propertyCallbacks = whenDefinedCallbacks.get(object).get(property);
-            propertyCallbacks.push(callback);
-            Object.defineProperty(object, property, {
-                configurable: true,
-                enumerable: true,
-                writeable: true,
-                /**
-                 * Retrieves the watched property from overloaded index
-                 *
-                 * @returns {*} {@see property} value on {@see object}
-                 */
-                get: function () {
-                    return this[overloadedPropertyName];
-                },
-                /**
-                 * Sets the overloaded property index with the provided value then executes configured callbacks
-                 *
-                 * @param {mixed} value
-                 */
-                set: function (value) {
-                    this[overloadedPropertyName] = value;
-                    for (var propertyCallback of propertyCallbacks.values()) {
-                        propertyCallback();
+                whenDefinedCallbacks.get(object).set(property, new Map([]));
+                Object.defineProperty(object, property, {
+                    configurable: true,
+                    enumerable: true,
+                    writeable: true,
+                    /**
+                     * Retrieves the watched property from overloaded index
+                     *
+                     * @returns {*} {@see property} value on {@see object}
+                     */
+                    get: function () {
+                        return this[overloadedPropertyName];
+                    },
+                    /**
+                     * Sets the overloaded property index with the provided value then executes configured callbacks
+                     *
+                     * @param {mixed} value
+                     */
+                    set: function (value) {
+                        this[overloadedPropertyName] = value;
+                        for (var propertyCallback of whenDefinedCallbacks.get(object).get(property).values()) {
+                            propertyCallback();
+                        }
                     }
-                }
-            });
+                });
+            }
+            if (typeof key == 'undefined') {
```

Comment:
Alternative option: if there are few calls with undefined we may add only the first callback

Reply:
This component is already used in the most complex parts of the code, changing the behavior like this can be unnecessarily risky